new(falco): add gVisor support

LucaGuerra opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. . Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> If contributing rules or changes to rules, please make sure to also uncomment one of the following line:

> /kind rule-update

> /kind rule-create

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area engine

> /area rules

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**What this PR does / why we need it**:

This PR adds support for the Falco CLI to enable syscall data collection from gVisor, as supported by libsinsp and libscap. Before this can be tested it needs the libs version to be advanced to the appropriate commit in Falco.

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` or `kind/flaky-test`, please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

See inline comments.

**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of the rule engine`.
-->

```release-note
new(falco): add gVisor support
```

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2078#" title="Author self-approved">LucaGuerra</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **leogr** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @leogr` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["leogr"]} -->

LucaGuerra commented on userspace/falco/application.h:
```diff
@@ -193,6 +193,11 @@ class application {
 
+	inline bool is_gvisor_enabled() const
+	{
+		return m_state->inspector->is_gvisor();
```
Note for the reviewer: the astute reader will have noticed that the boolean method `is_gvisor()` does not exist yet in falco libs, so I'm going to add it. I think it's the clean way to implement this check but if maintainers and contributors have better ideas I can change the implementation.

LucaGuerra replied:
Now it is implemented.

jasondellaluce commented:
/milestone 0.32.1

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2078#" title="Author self-approved">LucaGuerra</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **leodido** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @leodido` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["leodido"]} -->

FedeDP commented:
I think this should be rebased on master to fix integration tests!

LucaGuerra commented:
* PR is rebased
* It now requires https://github.com/falcosecurity/libs/pull/429 because the interface changed.

LucaGuerra commented:
Now it's rebased on top of libs that expose all the necessary functionality! :tada:

FedeDP commented on userspace/falco/app_cmdline_options.cpp:
```diff
@@ -162,6 +162,9 @@ void cmdline_options::define()
 		("disable-source",                "Disable a specific event source. Available event sources are: syscall or any source from a configured plugin with event sourcing capability. It can be passed multiple times. Can not disable all event sources.", cxxopts::value(disable_sources), "<event_source>")
 		("D",                             "Disable any rules with names having the substring <substring>. Can be specified multiple times. Can not be specified with -t.", cxxopts::value(disabled_rule_substrings), "<substring>")
 		("e",                             "Read the events from <events_file> in .scap format instead of tapping into live.", cxxopts::value(trace_filename), "<events_file>")
+		("g,gvisor-config",				  "Parse events from gVisor using the specified configuration file. A falco-compatible configuration file can be generated with --gvisor-generate-config and can be used for both runsc and Falco.", cxxopts::value(gvisor_config), "<gvisor_config>")
```
This LGTM! I am waiting for this part and we are ready to go IMO :)

leogr replied:
Use the `MINIMAL_BUILD` pattern please :point_down: 
https://github.com/falcosecurity/falco/blob/master/userspace/falco/app_cmdline_options.cpp#L166

Options must be present only when Falco has been built with the gVisor support.

Furthermore, note that the gVisor support must be disabled when `MINIMAL_BUILD` is ON.

LucaGuerra replied:
I'm following that same pattern, reverting the PR to wip until I'm relatively happy with the result

leogr commented on cmake/modules/falcosecurity-libs.cmake:
```diff
@@ -27,8 +27,8 @@ else()
   if(NOT FALCOSECURITY_LIBS_VERSION)
-    set(FALCOSECURITY_LIBS_VERSION "0.7.0-alpha2")
-    set(FALCOSECURITY_LIBS_CHECKSUM "SHA256=4a62071043aa50cdd29016c7dc041b401f8fa551271ec7b3ba01a537e5eb8bce")
+    set(FALCOSECURITY_LIBS_VERSION "a3b4b5bd586b491d329529d827d50443f98b240a")
```
I'd suggest tagging a version in libs before merging this PR.

@FedeDP wdyt?

FedeDP replied:
Makes sense! :+1:

leogr reviewed (commented):
This is awesome! :star_struck: 

Just left two comments (see below)

jasondellaluce commented on userspace/falco/app_actions/init_falco_engine.cpp:
```diff
@@ -43,6 +43,11 @@ void application::configure_output_format()
 	}
+	else if(m_options.gvisor_config != "")
+	{
+		output_format = "container=%container.id pid=%proc.vpid tid=%thread.vtid ";
```
```suggestion
		output_format = "container=%container.id pid=%proc.vpid tid=%thread.vtid";
```

jasondellaluce reviewed (commented):
Left you some comments Luca! Mostly nitpicking

LucaGuerra commented on userspace/falco/application.h:
```diff
@@ -193,6 +194,11 @@ class application {
 
+	inline bool is_gvisor_enabled() const
+	{
+		return m_state->inspector->is_gvisor();
```
Good catch! I'm not completely convinced of both but now I see that the one just above it does exactly what you describe. I updated that to reflect this.

LucaGuerra commented:
@leogr @jasondellaluce @FedeDP addressed review comments.
I also checked that it builds without gvisor support and in that case the CLI options related to it are not displayed at all.

FedeDP commented:
>  I also checked that it builds without gvisor support and in that case the CLI options related to it are not displayed at all.

Did you also check that gvisor is actually not built inside libscap? I mean, that eg: you don't see the protobuf dep in libscap.

LucaGuerra commented:
/retest

FedeDP commented:
IMO this LGTM! I think that we can wait for https://github.com/falcosecurity/libs/pull/433 and https://github.com/falcosecurity/libs/pull/416 in libs and tag a new libs/driver release that can possibly be the final for the release.
WDYT? 
Do you want this merged before those 2 PRs?

LucaGuerra commented:
@FedeDP  The PRs you mentioned are in and the versions have been updated in master. This can now be merged on top of master with the current RC libs versions. I just rebased the branch accordingly.

FedeDP reviewed (approved):
/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: 7606c85811917955e13d2213f78c499b7013e0a2</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2078#pullrequestreview-1026133948" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2078#" title="Author self-approved">LucaGuerra</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
